src/network/networkd-manager-varlink.c

```diff
@@ -56,6 +57,95 @@ static int vl_method_get_namespace_id(Varlink *link, JsonVariant *parameters, Va
+                { "InterfaceName",  JSON_VARIANT_STRING,        json_dispatch_const_string, offsetof(InterfaceInfo, ifname),  0 },
+                {}
+        };
+
+        InterfaceInfo info = {};
+        Link *link = NULL;
+        int r;
+
+        assert(vlink);
+        assert(manager);
+
+        r = varlink_dispatch(vlink, parameters, dispatch_table, &info);
+        if (r != 0)
+                return r;
+
+        if (info.ifindex < 0)
+                return varlink_error_invalid_parameter(vlink, JSON_VARIANT_STRING_CONST("InterfaceIndex"));
+        if (info.ifindex > 0) {
+                if (info.ifname)
+                        return varlink_error_invalid_parameter(vlink, JSON_VARIANT_STRING_CONST("InterfaceName"));
+
+                if (link_get_by_index(manager, info.ifindex, &link) < 0)
+                        return varlink_error_invalid_parameter(vlink, JSON_VARIANT_STRING_CONST("InterfaceIndex"));
+        }
+
+        if (info.ifname && link_get_by_name(manager, info.ifname, &link) < 0)
+                return varlink_error_invalid_parameter(vlink, JSON_VARIANT_STRING_CONST("InterfaceName"));
+
+        *ret = link;
+        return 0;
+}
+
+static int vl_method_get_lldp_neighbors(Varlink *vlink, JsonVariant *parameters, VarlinkMethodFlags flags, Manager *manager) {
+        Link *link = NULL;
+        int r;
+
+        assert(vlink);
+        assert(manager);
+
+        r = dispatch_interface(vlink, parameters, manager, &link);
+        if (r != 0)
+                return r;
+
+        if (link) {
+                _cleanup_(json_variant_unrefp) JsonVariant *w = NULL;
+
+                if (link->lldp_rx) {
+                        r = lldp_rx_build_neighbors_json(link->lldp_rx, &w);
+                        if (r < 0)
+                                return r;
+                }
+
+                return varlink_replyb(vlink, JSON_BUILD_OBJECT(JSON_BUILD_PAIR_VARIANT("Neighbors", w)));
+        }
+
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        HASHMAP_FOREACH(link, manager->links_by_index) {
+                _cleanup_(json_variant_unrefp) JsonVariant *w = NULL;
+
+                if (!link->lldp_rx)
+                        continue;
+
+                r = lldp_rx_build_neighbors_json(link->lldp_rx, &w);
+                if (r < 0)
+                        return r;
+
+                if (json_variant_is_blank_array(w))
+                        continue;
+
+                r = json_variant_append_arrayb(&v,
+                        JSON_BUILD_OBJECT(
+                                JSON_BUILD_PAIR_INTEGER("InterfaceIndex", link->ifindex),
+                                JSON_BUILD_PAIR_STRING("InterfaceName", link->ifname),
+                                JSON_BUILD_PAIR_VARIANT("Neighbors", w)));
+                if (r < 0)
+                        return r;
+        }
+
+        return varlink_replyb(vlink, JSON_BUILD_OBJECT(JSON_BUILD_PAIR_VARIANT("NeighborsByInterface", v)));
```

Comment:
I am not a fan of returning basically to distinct data structures here. Why not return the same array always, but once with a single entry, and once with many entries?

I think that would make things much nicer, and any specified ifindex/ifname would just be a filter, but not otherwise affect the structure of the data returned.